

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Common Arguments &mdash; Megalodon 2.3.3 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Advanced Megalodon Arguments" href="advanced_arguments.html" />
    <link rel="prev" title="Megalodon Algorithm Details" href="algorithm_details.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Megalodon
          

          
          </a>

          
            
            
              <div class="version">
                2.3.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="algorithm_details.html">Megalodon Algorithm Details</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Common Arguments</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#required-argument">Required Argument</a></li>
<li class="toctree-l2"><a class="reference internal" href="#guppy-backend-argument">Guppy Backend Argument</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-arguments">Output Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping-arguments">Mapping Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-variant-arguments">Sequence Variant Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#modified-base-arguments">Modified Base Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compute-resource-arguments">Compute Resource Arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advanced_arguments.html">Advanced Megalodon Arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="computing_considerations.html">Computing Considerations</a></li>
<li class="toctree-l1"><a class="reference internal" href="variant_phasing.html">Variant Phasing</a></li>
<li class="toctree-l1"><a class="reference internal" href="file_formats.html">File Formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_training.html">Megalodon Model Training</a></li>
<li class="toctree-l1"><a class="reference internal" href="modbase_training.html">Megalodon Modified Base Model Training</a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_aggregate.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">aggregate</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_calibrate.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">calibrate</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_merge.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">merge</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_modified_bases.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">modified_bases</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_phase_variants.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">phase_variants</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_per_read_text.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">per_read_text</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_validate.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">validate</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="extras_variants.html"><code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">variants</span></code></a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Megalodon</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Common Arguments</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/common_arguments.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="common-arguments">
<h1>Common Arguments<a class="headerlink" href="#common-arguments" title="Permalink to this headline">¶</a></h1>
<div class="section" id="required-argument">
<h2>Required Argument<a class="headerlink" href="#required-argument" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fast5s_dir</span></code></p>
<ul>
<li><p>Path to directory containing raw FAST5-format nanopore reads.</p></li>
<li><p>Both single and multi FAST5 formats are supported.</p></li>
<li><p>Default searches recursively for fast5 read files. To search only one-level specify <code class="docutils literal notranslate"><span class="pre">--not-recursive</span></code>.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="guppy-backend-argument">
<h2>Guppy Backend Argument<a class="headerlink" href="#guppy-backend-argument" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--guppy-config</span></code></p>
<ul>
<li><p>Guppy config.</p></li>
<li><p>Default: <code class="docutils literal notranslate"><span class="pre">dna_r9.4.1_450bps_modbases_5mc_hac.cfg</span></code></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--guppy-server-path</span></code></p>
<ul>
<li><p>Path to guppy server executable.</p></li>
<li><p>Default: <code class="docutils literal notranslate"><span class="pre">./ont-guppy/bin/guppy_basecall_server</span></code></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="output-arguments">
<h2>Output Arguments<a class="headerlink" href="#output-arguments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--live-processing</span></code></p>
<ul>
<li><p>As of version 2.2, Megalodon now supports live run processing.</p></li>
<li><p>Activate live processing mode by simply adding the <code class="docutils literal notranslate"><span class="pre">--live-processing</span></code> argument and specifying the MinKNOW output directory as the input FAST5 directory.</p></li>
<li><p>Megalodon will continue to search for FAST5s until the <code class="docutils literal notranslate"><span class="pre">final_summary*</span></code> file is created by MinKNOW, indicating data production has completed.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--outputs</span></code></p>
<ul>
<li><p>Specify desired outputs.</p></li>
<li><p>Options are <code class="docutils literal notranslate"><span class="pre">basecalls</span></code>, <code class="docutils literal notranslate"><span class="pre">mod_basecalls</span></code>, <code class="docutils literal notranslate"><span class="pre">mappings</span></code>, <code class="docutils literal notranslate"><span class="pre">variant_mappings</span></code>, <code class="docutils literal notranslate"><span class="pre">mod_mappings</span></code>, <code class="docutils literal notranslate"><span class="pre">per_read_variants</span></code>, <code class="docutils literal notranslate"><span class="pre">per_read_mods</span></code>, <code class="docutils literal notranslate"><span class="pre">variants</span></code>, and <code class="docutils literal notranslate"><span class="pre">mods</span></code>.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">mod_basecalls</span></code> are output in a BAM file via the <code class="docutils literal notranslate"><span class="pre">Mm</span></code> and <code class="docutils literal notranslate"><span class="pre">Ml</span></code> tags <a class="reference external" href="https://github.com/samtools/hts-specs/pull/418">described by hts-specs here</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">variant_mappings</span></code> are intended for obtaining highly accurate phased variant genotypes, but also provide a nice genome browser visualiztion of per-read variant calls.</p>
<ul>
<li><p>These mappings contain reference sequence at all positions except for per-read called variants. The base quality scores encode the likelihood for that reference anchored variant for use in the <code class="docutils literal notranslate"><span class="pre">whathap</span></code> phasing algorithm.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">mod_mappings</span></code> provide reference-anchored per-read modified base calls.</p>
<ul>
<li><p>As of version 2.2, the default output uses the <code class="docutils literal notranslate"><span class="pre">Mm</span></code> and <code class="docutils literal notranslate"><span class="pre">Ml</span></code> hts-specs tags (see above) with all modified bases in one output file.</p></li>
<li><p>Specify the <code class="docutils literal notranslate"><span class="pre">--mod-map-emulate-bisulfite</span></code> option to output one BAM per modified base with modified bases converted using <code class="docutils literal notranslate"><span class="pre">--mod-map-base-conv</span></code></p>
<ul>
<li><p>This file is useful for visualizing per-read modified base calls (e.g. IGV bisulfite mode for CpG calls).</p></li>
<li><p>This file may also allow a port to standard bisulfite pipelines that are capable of processing long-reads.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><p>Default output is <code class="docutils literal notranslate"><span class="pre">basecalls</span></code> only.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--output-directory</span></code></p>
<ul>
<li><p>Specify the directory to output results.
Default <code class="docutils literal notranslate"><span class="pre">megalodon_results</span></code></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--overwrite</span></code></p>
<ul>
<li><p>Overwrite the <code class="docutils literal notranslate"><span class="pre">--output-directory</span></code> if it exists.</p></li>
<li><p>Note that this is a recursive file deletion and should be used with caution.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="mapping-arguments">
<h2>Mapping Arguments<a class="headerlink" href="#mapping-arguments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--mappings-format</span></code></p>
<ul>
<li><p>Format for <code class="docutils literal notranslate"><span class="pre">mapping</span></code> output.</p></li>
<li><p>Options include <code class="docutils literal notranslate"><span class="pre">bam</span></code> (default), <code class="docutils literal notranslate"><span class="pre">cram</span></code>, and <code class="docutils literal notranslate"><span class="pre">sam</span></code>.</p></li>
<li><p>As of version 2.2, mappings are no longer sorted by default.</p>
<ul>
<li><p>Set <code class="docutils literal notranslate"><span class="pre">--sort-mappings</span></code> to sort mappings. If <code class="docutils literal notranslate"><span class="pre">samtools</span></code> is not in <code class="docutils literal notranslate"><span class="pre">$PATH</span></code> provide path to executable via the <code class="docutils literal notranslate"><span class="pre">--samtools-executable</span></code> argument.</p></li>
</ul>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--reference</span></code></p>
<ul>
<li><p>Reference genome or transcriptome in FASTA or minimap2 index format.</p>
<ul>
<li><p>If <code class="docutils literal notranslate"><span class="pre">--reference</span></code> is a minimap2 index and <code class="docutils literal notranslate"><span class="pre">--mapping-format</span></code> is <code class="docutils literal notranslate"><span class="pre">cram</span></code>, provide FASTA reference via <code class="docutils literal notranslate"><span class="pre">--cram-reference</span></code>.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="sequence-variant-arguments">
<h2>Sequence Variant Arguments<a class="headerlink" href="#sequence-variant-arguments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--haploid</span></code></p>
<ul>
<li><p>Compute sequence variants assuming a haploid reference. Default: diploid</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--variant-filename</span></code></p>
<ul>
<li><p>File containing putative variants in VCF/BCF format.</p>
<ul>
<li><p>Variants file must be sorted.</p></li>
<li><p>If variant file is not compressed and indexed this will be performed before further processing.</p></li>
</ul>
</li>
<li><p>Variants must be matched to the <code class="docutils literal notranslate"><span class="pre">--reference</span></code> provided.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="modified-base-arguments">
<h2>Modified Base Arguments<a class="headerlink" href="#modified-base-arguments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--mod-motif</span></code></p>
<ul>
<li><p>Restrict modified base results to the specified motifs.</p></li>
<li><p>This argument takes 3 values representing:</p>
<ol class="arabic simple">
<li><p>Modified base single letter codes (see <code class="docutils literal notranslate"><span class="pre">megalodon_extras</span> <span class="pre">modified_bases</span> <span class="pre">describe_alphabet</span></code> command)</p></li>
<li><p>Canonical sequence motif (may contain <a class="reference external" href="https://droog.gs.washington.edu/parc/images/iupac.html">ambiguity codes</a>)</p></li>
<li><p>Relative position (0-based) of the modified base within the canonical sequence motif</p></li>
</ol>
</li>
<li><p>Multiple <code class="docutils literal notranslate"><span class="pre">--mod-motif</span></code> arguments can be provided to a single <code class="docutils literal notranslate"><span class="pre">megalodon</span></code> command.</p></li>
<li><p>If not provided (and <code class="docutils literal notranslate"><span class="pre">per_read_mods</span></code> or <code class="docutils literal notranslate"><span class="pre">mods</span></code> outputs requested) all relevant sites are tested (e.g. all <code class="docutils literal notranslate"><span class="pre">C</span></code> bases for <code class="docutils literal notranslate"><span class="pre">5mC</span></code>).</p>
<ul>
<li><p>Note that restricting to motifs of interest can save computationally expensive steps and is considered more than a simple post-processing filter.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="compute-resource-arguments">
<h2>Compute Resource Arguments<a class="headerlink" href="#compute-resource-arguments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--processes</span></code></p>
<ul>
<li><p>Number of CPU read-processing workers to spawn.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">--devices</span></code></p>
<ul>
<li><p>GPU devices to use for basecalling acceleration.</p></li>
<li><p>If not provided CPU basecalling will be performed.</p></li>
<li><p>Device names can be provided in the following formats: <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">cuda0</span></code> or <code class="docutils literal notranslate"><span class="pre">cuda:0</span></code>.</p></li>
<li><p>Multiple devices can be specified separated by a space.</p></li>
</ul>
</li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="advanced_arguments.html" class="btn btn-neutral float-right" title="Advanced Megalodon Arguments" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="algorithm_details.html" class="btn btn-neutral float-left" title="Megalodon Algorithm Details" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2019, Oxford Nanopore Technologies.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>